### PR TITLE
[ADR] Sprint 17 Architecture Decision Records

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: src/adr/0026-removing-configured-additional-mailboxListeners.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+James enables a user to register additional mailbox listeners.
+
+The distributed James server is handling mailbox event processing (mailboxListener execution) using a RabbitMQ work-queue
+per listener.
+
+The distributed James server then declares a queue upon start for each one of these user registered listeners, that it
+binds to the main event exchange. 
+
+If the user unconfigures the listener, the queue and the binding are still present but not consumed. This result in
```
Comment (Arsnael) [inline src/adr/0026-removing-configured-additional-mailboxListeners.md:19]: s/result/results

File: src/adr/0026-removing-configured-additional-mailboxListeners.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+The distributed James server then declares a queue upon start for each one of these user registered listeners, that it
+binds to the main event exchange. 
+
+If the user unconfigures the listener, the queue and the binding are still present but not consumed. This result in 
+unbounded queue growth eventually causing RabbitMQ resource exhaustion and failure.
+
+## Decision
+
+We need, upon start, to sanitize bindings, and removes the ones corresponding to mailboxListeners that are not configured,
```
Comment (Arsnael) [inline src/adr/0026-removing-configured-additional-mailboxListeners.md:24]: s/removes/remove

File: src/adr/0026-removing-configured-additional-mailboxListeners.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+using a name matching strategy.
+
+The queue should not be deleted to prevent message loss.
+
+Given a James topology with a non uniform configuration, the effective RabbitMQ routing will be the one of the latest 
+started James server.
+
+## Alternatives
+
+We could also considering adding a webadmin endpoint to sanitize eventBus bindings, allowing more predictability than the
```
Comment (Arsnael) [inline src/adr/0026-removing-configured-additional-mailboxListeners.md:34]: s/considering/consider

File: src/adr/0026-removing-configured-additional-mailboxListeners.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+above solution but it would require admin intervention/
```
Comment (Arsnael) [inline src/adr/0026-removing-configured-additional-mailboxListeners.md:35]: the `\` at the end

File: src/adr/0027-eventBus-error-handling-upon-dispatch.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
+# 26. EventBus error handling upon dispatch
```
Comment (Arsnael) [inline src/adr/0027-eventBus-error-handling-upon-dispatch.md:1]: 27

File: src/adr/0028-Recompute-mailbox-quotas.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+## Status
+
+Accepted (lazy consensus)
+
+## Context
+
+JMAP custom quota extension, as well as IMAP [RFC-2087](https://tools.ietf.org/html/rfc2087) enables a user to monitor
+the amount of space and message count he is allowed to use, and that he is effectively using.
+
+To track the quota values a user is effectively using James relies on the
```
Comment (Arsnael) [inline src/adr/0028-Recompute-mailbox-quotas.md:14]: Something is wrong in this sentence... I think I would drop the `is`

File: src/adr/0028-Recompute-mailbox-quotas.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+Accepted (lazy consensus)
+
+## Context
+
+JMAP custom quota extension, as well as IMAP [RFC-2087](https://tools.ietf.org/html/rfc2087) enables a user to monitor
+the amount of space and message count he is allowed to use, and that he is effectively using.
+
+To track the quota values a user is effectively using James relies on the 
+[enventBus](../site/markdown/server/manage-guice-distributed-james.md#mailbox-event-bus) to increment a cassandra counter
```
Comment (Arsnael) [inline src/adr/0028-Recompute-mailbox-quotas.md:15]: s/cassandra/Cassandra

File: src/adr/0028-Recompute-mailbox-quotas.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+However, upon Cassandra failure, this value can be incorrect, hence the need of correcting it.
+
+## Data model details
+
+Table: imapUidTable: Holds mailbox and flags for each message, lookup by message ID
+
+Table: messageV2: Holds message metadata, independently of any mailboxes. Content of messages is stored in `blobs` 
+       and `blobparts` tables.
+       
+Table: currentQuota: Holds per quota-root current values. Quota-roots defines groups of mailboxes which shares quotas
```
Comment (Arsnael) [inline src/adr/0028-Recompute-mailbox-quotas.md:27]: s/shares/share ?

File: src/adr/0028-Recompute-mailbox-quotas.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+Table: imapUidTable: Holds mailbox and flags for each message, lookup by message ID
+
+Table: messageV2: Holds message metadata, independently of any mailboxes. Content of messages is stored in `blobs` 
+       and `blobparts` tables.
+       
+Table: currentQuota: Holds per quota-root current values. Quota-roots defines groups of mailboxes which shares quotas 
+limitations.
+
+Operation:
+ - Quota updates is done asynchronously (event bus + listener) for successful mailbox operations.
```
Comment (Arsnael) [inline src/adr/0028-Recompute-mailbox-quotas.md:31]: s/is/are

File: src/adr/0028-Recompute-mailbox-quotas.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+   - If the quota update is not applied, then we are inconsistent
+   - To be noted, event bus mechanisms should already limit
```
Comment (Arsnael) [inline src/adr/0028-Recompute-mailbox-quotas.md:33]: should already be limited ?

File: src/adr/0025-cassandra-blob-store-cache.md [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+to be highly effective. We expect comparable performance improvments compared to an uncache ObjectStorage blob store.
+
+HybridBlobStore should be removed.
+
+## Alternatives
+
+[14. Add storage policies for BlobStore](0014-blobstore-storage-policies.md) proposes to use the CassandraBlobStore to
+mimic a cache.
+
+This solution needed further work as we decided to add an option to write all blobs to the object storage in order:
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply src/adr/0025-cassandra-blob-store-cache.md:56]: Passed tense intended as `HybridBlobStore should be removed.` above

File: src/adr/0028-Recompute-mailbox-quotas.md (hunk 6 above)
Comment (chibenwa) [reply src/adr/0028-Recompute-mailbox-quotas.md:14]: More a coma

File: src/adr/0025-cassandra-blob-store-cache.md [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+James exposes a simple BlobStore API for storing raw data. However such raw data often vary in size and access patterns.
+
+As an example:
+
+ - Mailbox message headers are expected to be small and frequently accessed
+ - Mailbox message body are expected to have sizes ranging from small to big but are unfrequently accessed
+ - DeletedMessageVault message headers are expected to be small and unfrequently accessed
+
+The access pattern of some of these kind of blobs do not fit Object Storage caracteristics: good at storing big blobs, but
```
Comment (mbaechler) [inline src/adr/0025-cassandra-blob-store-cache.md]: you could define what is "slow". I think observed latency is around 50ms vs 1ms expected from cassandra

File: src/adr/0025-cassandra-blob-store-cache.md [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+ - DeletedMessageVault message headers are expected to be small and unfrequently accessed
+
+The access pattern of some of these kind of blobs do not fit Object Storage caracteristics: good at storing big blobs, but 
+it induces high latencies for reading small blobs.
+
+This get's some operations slow (for instance IMAP FETCH headers, or listing JMAP messages).
+
+## Decision
+
+We need to implement a write through cache to decrease the amount of object read from object storage.
```
Comment (mbaechler) [inline src/adr/0025-cassandra-blob-store-cache.md]: please remove "need to".

Comment (mbaechler) [inline src/adr/0025-cassandra-blob-store-cache.md]: > decrease the amount of object read from object storage

have better read latency for smaller objects

File: src/adr/0025-cassandra-blob-store-cache.md [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+Such a cache needs to be distributed in order to be more efficient.
+
+Given that we don't want to introduce new technologies, we will implement it using Cassandra.
+
+The cache should be implemented as a key-value table on a dedicated 'cache' keyspace, with a reprication factor of 1, 
+and be queried with a consistency level of ONE. 
+
+We will leverage a configurable TTL as an eviction policy. Cache will be populated upon writes and missed read, if the
```
Comment (mbaechler) [inline src/adr/0025-cassandra-blob-store-cache.md:37]: what compaction will we use?

File: src/adr/0025-cassandra-blob-store-cache.md [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+[14. Add storage policies for BlobStore](0014-blobstore-storage-policies.md) proposes to use the CassandraBlobStore to
+mimic a cache.
+
+This solution needed further work as we decided to add an option to write all blobs to the object storage in order:
+ - To get a centralized source of truth
+ - Being able to instantly rollback Hybrid blob store adotpion
+ 
+See [this pull request](https://github.com/linagora/james-project/pull/3162)
+
+With such a proposal there is no eviction policy. Also, the storage is done on the main keysapce with a high replication
```
Comment (mbaechler) [inline src/adr/0025-cassandra-blob-store-cache.md]: s/keysapce/keyspace

File: src/adr/0026-removing-configured-additional-mailboxListeners.md (hunk 3 above)
Comment (mbaechler) [inline src/adr/0026-removing-configured-additional-mailboxListeners.md]: what about using EventSourcing to handle configuration changes regarding listeners?

File: src/adr/0027-eventBus-error-handling-upon-dispatch.md [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+James allows asynchronous processing for mailbox events via MailboxListener. This processing is abstracted by the 
+EventBus.
+
+If the processing of an event via a mailbox listener fails, it is retried, until it succeeds. If a maxRetries parameter 
+is exceeded, the event is stored in deadLetter and no further processing is attended.
+
+The administrator can then look at the content of deadLetter to diagnose processing issues and schedule a reDelivery in 
+order to retry their processing via webAdmin APIs.
+
+However no such capabilities are supported upon dispatch. A failed dispatch will result in message loss.
```
Comment (mbaechler) [inline src/adr/0027-eventBus-error-handling-upon-dispatch.md:20]: what do you mean by "dispatch" ?

File: src/adr/0028-Recompute-mailbox-quotas.md [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+
+Accepted (lazy consensus)
+
+## Context
+
+JMAP custom quota extension, as well as IMAP [RFC-2087](https://tools.ietf.org/html/rfc2087) enables a user to monitor
+the amount of space and message count he is allowed to use, and that he is effectively using.
+
+To track the quota values a user is effectively using, James relies on the 
+[enventBus](../site/markdown/server/manage-guice-distributed-james.md#mailbox-event-bus) to increment a Cassandra counter
```
Comment (remk) [inline src/adr/0028-Recompute-mailbox-quotas.md:15]: s/enventBus/eventBus

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**